adam_core/utils/indexable.py

```diff
@@ -1,174 +1,446 @@
-                self.__dict__[k] = Time(
-                    np.concatenate([self_v.mjd, v.mjd]), scale=v.scale, format="mjd"
+        # --- Check boundaries on the slice
+        if isinstance(ind, slice) and ind.start is not None and ind.start >= len(self):
+            raise IndexError(f"Index {ind.start} is out of bounds.")
+
+        if isinstance(ind, int) and self._class_index_to_members_is_slice:
+            member_ind = self._class_index_to_members[ind]
+
+        elif isinstance(ind, slice) and self._class_index_to_members_is_slice:
+
+            # Check if the array of slices are consecutive and share
+            # the same step size. If so, create a single slice that
+            # combines all of the slices.
+            slices = self._class_index_to_members[ind]
+            is_consecutive = True
+            for i, s_i in enumerate(slices[:-1]):
+                if s_i.stop != slices[i + 1].start:
+                    is_consecutive = False
+                    break
+                if s_i.step is not None and (s_i.step != slices[i + 1].step):
+                    is_consecutive = False
+                    break
+
+            if is_consecutive:
+                logger.debug(
+                    "Slices are consecutive and share the same step. "
+                    f"Combining slices a single slice with start {slices[0].start}, "
+                    f"end {slices[-1].stop} and step {slices[0].step}."
+                )
+                member_ind = slice(slices[0].start, slices[-1].stop, slices[0].step)
+            else:
+                logger.debug(
+                    "Slices are not consecutive. "
+                    "Combining slices a concatenating the members index for each slice."
+                )
+                member_ind = np.concatenate([self._class_index[s] for s in slices])
+
+        elif np.array_equal(self._class_index, self._member_index):
+            logger.debug("Using class index to index member arrays.")
+            member_ind = self._class_index[ind]
+        else:
+            logger.debug(
+                "Using unique class index to index member arrays with np.isin."
+            )
+            member_ind = self._member_index[
+                np.isin(
+                    self._class_index_to_members,
+                    self._class_index[ind],
                 )
+            ]
+        return member_ind
 
-            elif isinstance(v, (list, Indexable)):
-                self_v.append(v)
+    def __len__(self):
+        if self._class_index is not None:
+            return len(self._class_index)
+        else:
+            err = "Length is not defined for this class."
+            raise NotImplementedError(err)
 
-            elif isinstance(v, pd.MultiIndex):
-                self.set_index(index=self._index_attribute)
+    def __getitem__(self, class_ind: Union[int, slice, list, np.ndarray]):
+
+        member_ind = self.query_index(class_ind)
+        copy_self = copy(self)
```

Comment:
It is weird for sure. And I'm trying to think of a use case where I wouldn't want to access the original arrays rather than work on a copy of them.. and I can't think of an example. 

What do you think of this? 
```
    def __getitem__(self, class_ind: Union[int, slice, list, np.ndarray]):

        member_ind = self._query_index(class_ind)
        item = self.__class__()

        for k, v in self.__dict__.items():
            if k != "_class_index":
                if isinstance(v, (np.ndarray, np.ma.masked_array, Time, Indexable)):
                    item.__dict__[k] = v[member_ind]
                elif isinstance(v, UNSLICEABLE_DATA_STRUCTURES):
                    item.__dict__[k] = v
                elif v is None:
                    pass
                else:
                    err = f"{type(v)} are not supported."
                    raise NotImplementedError(err)
            else:
                item.__dict__[k] = v[np.s_[class_ind]]

        return item
```

Reply:
That looks scary to me. I don't think you can get away with `self.__class__()`. The class's initializer could easily take arguments, and you don't know what arguments to pass in at that spot. I think `copy` is preferable over that.

This might be something to optimize once data is stored in a separate attribute rather than bundled into `__dict__`. Then we could control the data structure used, so copying (or granting a view on a subslice) could be made efficient.

Reply:
Understood, I've left it as a copy.